Py3

kvaml opened the pull request:
I have finished your tutorial and made the necessary changes to run under Python3.

The zope interfaces implements function has been replaced by an implementer decorator.

The network streams all use bytes (b'abcdef'). This required explicit conversions between bytes and strings which are now unicode. I added an italian poem to the poetry directory to make sure success was not depending upon ASCII data streams.

I think your tutorials hold up pretty well, but the code snippets are now slightly off.

Let me know if you want some additional changes and cleanup. Everything did work for me.

kvaml committed:
Gradually converting the Python programs to Python 3.

kvaml committed:
ignore the backup files from 2to3 conversion
these are now python 3 programs

kvaml committed:
Besides the usual 2to3 conversions also convert the screen position
computations to use integer (//) division.

kvaml committed:
Mostly routine 2to3 changes. Items to note:
The conversion from bytes to unicode needs to be managed explicitly now.
Ascii poems did not require any thought as to when to convert to unicode.
I preserved the global except statements, but I would urge you to get rid
of them. They should never be used casually because of the risk of "handling"
an exception that is totally unexpected and not really getting handled.

kvaml committed:
Removed the bytes.encode('utf8'). This proved to be an error when I added an Italian poem.
The file was read as a stream of bytes, so there was nothing to be done.
The file is UTF-8, so the decoding by the client application works correctly.

kvaml committed:
Added an Italian poem to better test the unicode handling.
Most changes relate to sending bytes over the network connections
and converting the byte streams to unicode strings.
Made changes to error handling to see errors that were getting suppressed.
Imported the implementer decorator to replace implements from zope.interfaces

kvaml committed:
Update README

kvaml committed:
Ususal changes from 2to3-3.5 updating print, etc.
Usual changes to use bytes for network data.

kvaml committed:
Merge branch 'py3' of https://github.com/kvaml/twisted-intro into py3

Merges changers made to README on the github web site.

jdavisp3 commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 twistd.log*
 twistd.pid
 *.pyc
+ex-*.py
```
What's this ignore for?

kvaml replied:
I did write some code for your exercises at the end of each lesson. Those programs started with ex-. At this point, I could just delete those files and revert the .gitignore.

jdavisp3 replied:
I see, the `ex` stands for "exercise". That makes sense now. Why don't we add a comment here explaining the purpose.

jdavisp3 commented on README:
```diff
@@ -3,4 +3,9 @@ and asynchronous programming in Python located here:
+future versions of Twisted will be Python 3 only.
+
+Originally by Dave Peticolas <dave@krondo.com>
+Being converted by lkvam@venix.com
```
```suggestion
Converted to Python 3 by lkvam@venix.com
```

jdavisp3 commented on poetry/alla_sera.txt:
```diff
@@ -0,0 +1,22 @@
+quello spirto guerrier ch’entro mi rugge.
+
+
+by Ugo Foscolo
```
👌

jdavisp3 commented on twisted-server-1/transformedpoetry.py:
```diff
@@ -45,20 +45,37 @@ def cummingsify(self, poem):
+        try:
+            new_poem = self.factory.transform(xform_name, poem)
+        except ValueError as exc:
+            self.sendString(str(exc))
```
I'm not sure I agree with this change. Errors should be clearly distinguished from correct return values. This silently returns an error as if it were the result of the transform.

kvaml replied:
OK. I'll review my changes and find a better solution. Some of the original error handling was a bit too optimistic since now even a print statement is fraught with the risk of a unicode error. The original ASCII poems were risk free in Python 2.

kvaml replied:
I pushed the changes.

The server now logs errors that had been silently swallowed before. The protocol works as specified; only poems are sent.

jdavisp3 commented on twisted/plugins/fastpoetry_plugin.py:
```diff
@@ -53,9 +53,12 @@ class Options(usage.Options):
-    implements(service.IServiceMaker, IPlugin)
+    # implements does not work in Python3
+    # see the implementer decorator above
+    ##implements(service.IServiceMaker, IPlugin)
```
Let's just delete these lines then, no need to leave them here.

kvaml replied:
Done.

jdavisp3 reviewed (commented):
Thanks for all this work!

kvaml committed:
Update README

Co-authored-by: Dave Peticolas <dave@krondo.com>

kvaml committed:
removed the comments in fastpoetry_plugin that referenced the obsolete implements function.

changed the error handling in the transformedpoetry server. I had changed the server to
send error messages in place of poems when errors occurred. That violates the protocol.
Now the server logs all of its errors. That should be sufficient to debug client and server
errors in the future.

kvaml committed:
Should test before push. I introduced a syntax error in my except statement in my previous
commit. That is now fixed and I log the error that triggered the exception.

kvaml committed:
I had never run this failure program. I may have skipped it as unnecessary for
my learning.
The only changes are the automatic ones from 2to3.

kvaml committed:
I think this is the last file I failed to convert.
The changes are the usual 2to3 conversions plus bytes for network streams
and conversion to unicode strings for finished poems.

jdavisp3 commented on blocking-client/get-poetry.py:
```diff
@@ -65,7 +65,8 @@ def get_poetry(address):
             break
 
-        poem += data
+        # convert network stream to unicode for poem
```
I think this comment is not needed, it's clear from the code what's happening.

kvaml replied:
OK. pushed change.

jdavisp3 commented on twisted-client-1/get-poetry.py:
```diff
@@ -86,7 +86,8 @@ def connectionLost(self, reason):
 
     def doRead(self):
-        bytes = ''
+        # data from socket is bytes
```
I think we should rename the variable so that it doesn't clash with the `bytes` built-in. And no need for the comment, the `b'` prefix makes that clear.

kvaml replied:
Renamed to chunks. Removed comment. Pushed change.

jdavisp3 commented on twisted-client-1/get-poetry.py:
```diff
@@ -95,19 +96,20 @@ def doRead(self):
+            print(msg % (self.task_num, len(bytes), self.format_addr()))
 
-        self.poem += bytes
+        #poem is unicode
```
Comment not needed.

kvaml replied:
removed and pushed.

jdavisp3 commented on twisted-client-2/get-poetry-simple.py:
```diff
@@ -73,7 +73,10 @@ def __init__(self, poetry_count):
-            self.poems.append(poem)
+            # this seems like the place to convert poem from bytes to unicode
+            # the network code dealt with bytes
+            # but the completed poem should be a unicode string.
```
I don't think this comment is needed, I agree with putting the conversion here.

kvaml replied:
removed and pushed.

kvaml committed:
Removed unnecessary comments about bytes.
Changed variable name bytes to chunks. This avoids masking the bytes function/type.

jdavisp3 reviewed (approved):
@kvaml 🚀 Thanks so much for this, it's a big help!
